[bug] AutotoolsDeps does not correctly handle "shared-library" package types on -rpath linkage

### Describe the bug

From: https://github.com/conan-io/conan/blob/develop2/conan/tools/gnu/autotoolsdeps.py?plain=1#L29-L33

```py
    def _rpaths_flags(self):
        flags = []
        for dep in self.ordered_deps:
            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in dep.cpp_info.libdirs
                          if dep.options.get_safe("shared", False))
```

`AutotoolsDeps` only checks if the `shared` option exists in the dependency. But it happens that a dependency could be a `shared-library`, and `shared-library`s does not contain `shared` option. 

Easy workarord:
```diff
diff --git a/conan/tools/gnu/autotoolsdeps.py b/conan/tools/gnu/autotoolsdeps.py
index 8976c600e..e8dd42be0 100644
--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -30,7 +30,8 @@ class AutotoolsDeps:
         flags = []
         for dep in self.ordered_deps:
             flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in dep.cpp_info.libdirs
-                          if dep.options.get_safe("shared", False)])
+                          if dep.options.get_safe("shared", False)
+                          or dep.package_type == "shared-library"])
         return flags
 
     @property
```

### How to reproduce it

This bug was located while trying to compile `cyrus-sasl/2.1.28` with `with_gssapi=True`.

To correctly reproduce this error, `fix_apple_shared_install_name(self)` should be called from the `package()` method of `krb5` recipe. An incoming PR will be open soon.
 

```sh
$ conan cci:create recipes/cyrus-sasl -o '&:with_gssapi=True'
...
checking whether we are cross compiling... configure: error: in `/Users/perseo/.conan2/p/b/cyrus06ec5650ee044/b/build-release':                                                       
configure: error: cannot run C compiled programs.                                                                                                                                     
If you meant to cross compile, use `--host'.                                                                                                                                          
See `config.log' for more details                                                                                                                                                     
                                                                                                                                                                                      
cyrus-sasl/2.1.28: ERROR:                                                                                                                                                             
Package 'c54a3ef084c40103860ccf90c6a618b8b103edc0' build failed                                                                                                                       
cyrus-sasl/2.1.28: WARN: Build folder /Users/perseo/.conan2/p/b/cyrus06ec5650ee044/b/build-release                                                                                    
ERROR: cyrus-sasl/2.1.28: Error in build() method, line 232                                                                                                                           
        self._build_autotools()                                                                                                                                                       
while calling '_build_autotools', line 167                                                                                                                                            
        autotools.configure()                                                                                                                                                         
        ConanException: Error 1 while executing  
```

Autotools config.log contents:
```
dyld[24835]: Library not loaded: @rpath/libkadm5srv_mit.12.0.dylib
  Referenced from: <78EAC1CC-5311-39C1-A510-25BA8CA85C31> /Users/perseo/.conan2/p/b/cyrus06ec5650ee044/b/build-release/conftest
  Reason: no LC_RPATH's found
```

--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -29,7 +29,7 @@ class AutotoolsDeps:
     def _rpaths_flags(self):
         flags = []
         for dep in self.ordered_deps:
-            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in dep.cpp_info.libdirs
+            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in self._get_cpp_info().libdirs
                           if dep.package_type == "shared-library"])
         return flags
 

--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -30,9 +30,9 @@ class AutotoolsDeps:
     def _rpaths_flags(self):
         flags = []
         for dep in self.ordered_deps:
-            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir)
-                          for libdir in dep.cpp_info.aggregated_components().libdirs
-                          if dep.package_type is PackageType.SHARED])
+            if dep.package_type is PackageType.SHARED:
+                flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in
+                              dep.cpp_info.aggregated_components().libdirs])
         return flags
 
     @property

--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -29,7 +29,8 @@ class AutotoolsDeps:
     def _rpaths_flags(self):
         flags = []
         for dep in self.ordered_deps:
-            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in self._get_cpp_info().libdirs
+            flags.extend(["-Wl,-rpath -Wl,{}".format(libdir)
+                          for libdir in dep.cpp_info.aggregated_components().libdirs
                           if dep.package_type == "shared-library"])
         return flags
 
@@ -79,6 +80,6 @@ class AutotoolsDeps:
     def vars(self, scope="build"):
         return self.environment.vars(self._conanfile, scope=scope)
 
-    def generate(self,  scope="build"):
+    def generate(self, scope="build"):
         check_duplicated_generator(self, self._conanfile)
         self.vars(scope).save_script("conanautotoolsdeps")

--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -1,4 +1,5 @@
 from conan.internal import check_duplicated_generator
+from conan.internal.model.pkg_type import PackageType
 from conan.tools import CppInfo
 from conan.tools.env import Environment
 from conan.tools.gnu.gnudeps_flags import GnuDepsFlags
@@ -31,7 +32,7 @@ class AutotoolsDeps:
         for dep in self.ordered_deps:
             flags.extend(["-Wl,-rpath -Wl,{}".format(libdir)
                           for libdir in dep.cpp_info.aggregated_components().libdirs
-                          if dep.package_type == "shared-library"])
+                          if dep.package_type is PackageType.SHARED])
         return flags
 
     @property

--- a/conan/tools/gnu/autotoolsdeps.py
+++ b/conan/tools/gnu/autotoolsdeps.py
@@ -30,7 +30,7 @@ class AutotoolsDeps:
         flags = []
         for dep in self.ordered_deps:
             flags.extend(["-Wl,-rpath -Wl,{}".format(libdir) for libdir in dep.cpp_info.libdirs
-                          if dep.options.get_safe("shared", False)])
+                          if dep.package_type == "shared-library"])
         return flags
 
     @property

--- a/test/integration/toolchains/gnu/test_autotoolsdeps.py
+++ b/test/integration/toolchains/gnu/test_autotoolsdeps.py
@@ -155,10 +155,11 @@ def test_autotoolsdeps_macos_rpath_by_package_type(package_type, use_shared_opti
             "conanfile.py": consumer,
             "macos": profile})
     t.run("create dep_pkg --profile:host=macos")
+    dp_pkg_package_path = t.created_layout().package()
     t.run("install --profile:host=macos")
     deps = t.load("conanautotoolsdeps.sh")
     if expect_rpath:
-        assert "-Wl,-rpath" in deps
+        assert f"-Wl,-rpath -Wl,{dp_pkg_package_path}" in deps
     else:
         assert "-Wl,-rpath" not in deps
 
@@ -194,7 +195,8 @@ def test_autotoolsdeps_macos_rpath_shared_dep_with_components():
             "conanfile.py": consumer,
             "macos": profile})
     t.run("create dep_pkg --profile:host=macos")
+    dp_pkg_package_path = t.created_layout().package()
     t.run("install --profile:host=macos")
     deps = t.load("conanautotoolsdeps.sh")
     assert "nested" in deps
-    assert "-Wl,-rpath" in deps
+    assert f"-Wl,-rpath -Wl,{dp_pkg_package_path}" in deps

--- a/test/integration/toolchains/gnu/test_autotoolsdeps.py
+++ b/test/integration/toolchains/gnu/test_autotoolsdeps.py
@@ -155,9 +155,46 @@ def test_autotoolsdeps_macos_rpath_by_package_type(package_type, use_shared_opti
             "conanfile.py": consumer,
             "macos": profile})
     t.run("create dep_pkg --profile:host=macos")
-    t.run("install . --profile:host=macos")
+    t.run("install --profile:host=macos")
     deps = t.load("conanautotoolsdeps.sh")
     if expect_rpath:
         assert "-Wl,-rpath" in deps
     else:
         assert "-Wl,-rpath" not in deps
+
+
+@pytest.mark.skipif(platform.system() not in ["Linux", "Darwin"], reason="Autotools")
+def test_autotoolsdeps_macos_rpath_shared_dep_with_components():
+    profile = textwrap.dedent("""
+         [settings]
+         build_type=Release
+         arch=x86
+         os=Macos
+         compiler=gcc
+         compiler.libcxx=libstdc++11
+         compiler.version=7.1
+         compiler.cppstd=17
+    """)
+    dep = (GenConanfile("pkg", "1.0")
+           .with_settings("os", "arch", "compiler", "build_type")
+           .with_package_type("shared-library")
+           .with_package_info({
+               "libdirs": [],
+               "components": {
+                   "cmp": {"libdirs": ["nested/lib"], "libs": ["pkg"]},
+               },
+           }))
+    consumer = (GenConanfile("app", "1.0")
+                .with_settings("os", "arch", "compiler", "build_type")
+                .with_require("pkg/1.0")
+                .with_generator("AutotoolsDeps"))
+
+    t = TestClient()
+    t.save({"dep_pkg/conanfile.py": dep,
+            "conanfile.py": consumer,
+            "macos": profile})
+    t.run("create dep_pkg --profile:host=macos")
+    t.run("install --profile:host=macos")
+    deps = t.load("conanautotoolsdeps.sh")
+    assert "nested" in deps
+    assert "-Wl,-rpath" in deps

--- a/test/integration/toolchains/gnu/test_autotoolsdeps.py
+++ b/test/integration/toolchains/gnu/test_autotoolsdeps.py
@@ -116,3 +116,48 @@ def test_cpp_info_aggregation():
 
     t.save({"conanfile.py": consumer})
     t.run("create . --name consumer --version 1.0 --profile:host=macos")
+
+
+@pytest.mark.skipif(platform.system() not in ["Linux", "Darwin"], reason="Autotools")
+@pytest.mark.parametrize(
+    "package_type,use_shared_option,expect_rpath",
+    [
+        ("shared-library", False, True),
+        ("static-library", False, False),
+        ("library", True, True),
+    ],
+)
+def test_autotoolsdeps_macos_rpath_by_package_type(package_type, use_shared_option, expect_rpath):
+    """On Macos, ``AutotoolsDeps`` adds ``-Wl,-rpath`` only for host deps whose type is shared."""
+    profile = textwrap.dedent("""
+         [settings]
+         build_type=Release
+         arch=x86
+         os=Macos
+         compiler=gcc
+         compiler.libcxx=libstdc++11
+         compiler.version=7.1
+         compiler.cppstd=17
+    """)
+    dep = (GenConanfile("pkg", "1.0")
+           .with_settings("os", "arch", "compiler", "build_type")
+           .with_package_type(package_type)
+           .with_package_info({"libdirs": ["lib"], "libs": ["pkg"]}))
+    if use_shared_option:
+        dep = dep.with_shared_option(True)
+    consumer = (GenConanfile("app", "1.0")
+                .with_settings("os", "arch", "compiler", "build_type")
+                .with_require("pkg/1.0")
+                .with_generator("AutotoolsDeps"))
+
+    t = TestClient()
+    t.save({"dep_pkg/conanfile.py": dep,
+            "conanfile.py": consumer,
+            "macos": profile})
+    t.run("create dep_pkg --profile:host=macos")
+    t.run("install . --profile:host=macos")
+    deps = t.load("conanautotoolsdeps.sh")
+    if expect_rpath:
+        assert "-Wl,-rpath" in deps
+    else:
+        assert "-Wl,-rpath" not in deps

